### ISSUE TITLE
No straightforward, runnable examples

### ISSUE THREAD

**Issue by dsoprea** [hanwen/go-fuse#304]
This is a dignified rant (done with all due respect) about the lack of functional/developed (read: minimally complete) examples with this project, which otherwise seems mature and awesome. 

Just saying, "The examples aren't good enough.", at best, doesn't really express the difficulty of trying to implement all of FUSE's tiny details once they start piling up and, at worst, seems petty and subjective. The only way to effectively make the case for better examples is to share the process of trying to research the standard implementation details. FUSE is traditionally very convention-based. Generally, one good example using data stored in memory in an uncomplicated way is enough to give anyone enough information to proceed with whatever their requirement is. However, when the examples are partial implementations or implement alternative flows (perhaps used by certain formats that might require that), a treatment of the common use-case gets starved. I believe there's also the opportunity for multiple examples reusing a common implementation, which also potentially happens to be fairly twisted in its details, to compromise the effective value of all of those examples.

I currently need to implement a _Close()_ method for individual file-handles. So, this is the progression of my first steps to try to avoid having to delve down into the mechanics. I'm simply looking for a "Close()" implementation in one of the examples. So, I do the natural thing and revisit the hello example, but only see some weird _OnAdd_ method and a _Getattr()_ method (and think, "what is this supposed to be an example of?"). I then look at the _loopback_ example, but there's really just a _main()_ function and no FUSE functions at all (and think, "ok; is this to support some unit-test somewhere?"). I check the _memfs_ example, and it's, once more, mostly empty but for a _main()_ function. The same thing with _zipfs_. Not only am I having trouble finding methods that I'd like example of how to implement in this project, but most of the examples don't even implement any actual filesystems. They're just providing command-line or profiling logic and/or calling stuff in the internals of the project rather than serving as tutorials/examples.

I then grep for "Close(" in the examples directory (which has nine examples in it). Not only is there not one minimally complete example, but it seems like several of the basic, mandatory methods (e.g. Read, Write, Readdir, Flush, Open, Close; note that _Read()_ will not succeed without _Flush()_, at least in the _nodefs_ methodology of this project, so I would expect to find _Read()_ as well as _Flush()_, by association, in at least one of the examples) are totally missing from among the entire body of examples.

Ok. So, I go further down into the project. Looking into _zipfs/zipfs.go_ at the _zipFile_ type, it implements _Open()_, _Read()_, and _Getattr()_, but then defers to default/unimplemented methods for the rest. I look at the _tarfs.go_ type, but it only implements the go-fuse-specific _OnAdd()_ method.

Going into _posixtest/_, I see a bunch of convenience functions that don't appear to satisfy any of the FUSE interfaces. Going into _unionfs/_, I see _unionfs.go_, which appears to implement a lot of standard FUSE API methods. Not "Close(", though. So, I grep the directory for "Close(", and, still, nothing implements it.

All of this to say that, for people who are coming from an existing FUSE background, there is nothing overly helpful to start from. The people who aren't coming from a FUSE background will be totally confused because none of the examples bear any resemblance to the typical VFS API or standard FUSE implementations.

In order to implement a simple memory-based listing/reading implementation, I needed to do a lot of tracing and add a lot of verbosity when I would've expected to just be able to run an example and see some output, which I could then hack for my own purposes. I've since written a FUSE interface to a non-public archival format, but I'm also still having to dive into the mechanics rather than simply being able to cross-reference examples.

TLDR:

Can I submit a pull-request for my lister/reader example (https://github.com/dsoprea/go-fuse-example) in order to soften the curve for new implementors as well as to provide a reference-implementation in general? After I implement the writing support, I can circle-back to update the example to include that, too.

Thanks for your patience in reading this.

**Comment by hanwen** [comment]
A couple of comments: 

1. close is neither part of the kernel VFS nor, the FUSE API surface. I think you are looking for flush, 

https://kernel.googlesource.com/pub/scm/linux/kernel/git/torvalds/linux/+/d7d170a8e357bd9926cc6bfea5c2385c2eac65b2/include/linux/fs.h#1805
https://github.com/libfuse/libfuse/blob/249942f6411042c4af18bd10438da34801cce02b/include/fuse.h#L496

If you had looked for "close()" in the documentation, you would have probably found https://godoc.org/github.com/hanwen/go-fuse/fs#NodeFlusher

I'm adding some doc notes to the nodefs and pathfs packages to declaring them deprecated, pointing to the fs package.

Regarding the examples' apparent lack of logic: the examples are using file systems implemented in libraries, such as [NewLoopbackRoot](https://godoc.org/github.com/hanwen/go-fuse/fs#NewLoopbackRoot) and [NewZipTree](https://godoc.org/github.com/hanwen/go-fuse/zipfs#NewZipTree). This is so the code in use will be thoroughly unittested and kept up to date through API changes. Similarly, the [example](https://godoc.org/github.com/hanwen/go-fuse/fs#example-package) in the documentation uses MemRegularFile so it is built on tested, proven to work components.

I agree that the package could do with a better example of a read/write filesystem, but apart from the loopback file system, I can't think of an obvious example. It's typically also much harder to get correct, because of the inherent parallelism in the linux vfs layer. I'm happy to hear your suggestions.

**Comment by ghz-gehtz** [comment]
My 2ct for having used hanwen/go-fuse, jacobsa/fuse and billziss-gh/cgofuse for my project.

You should just make the loopback implementation more obvious as the best start for beginners for being a complete implementation of the API. It took me a while to figured out, especially if it's your first time having to deal with fuse and abstracted filesystem layer.

**Comment by hanwen** [comment]
@ghz-gehtz - do you mean that beginners should start reading the loopback implementation, or do you mean that beginners should use the loopback implementation as a building block through embedding?

**Comment by ghz-gehtz** [comment]
@hanwen for me, both.

**Comment by hanwen** [comment]
i've added an example for readdir and lookup. 

https://review.gerrithub.io/c/hanwen/go-fuse/+/462412/1/fs/dynamic_example_test.go

**Comment by hanwen** [comment]
i'd be happy for someone to look this over before I merge it.

**Comment by dsoprea** [comment]
Please add the standard documentation for the methods/symbols. Please add the expected output for the example per the [testable example](https://blog.golang.org/examples) requirements ("Output:"). Please add a reference to this document from the README and maybe a description/overview, somewhere, of what this is supposed to be an example of. 

This is all in the interest of helping new users as well as providing the minimum documentation that Go's documentation system prescribes, and, ergo, what most Go developers know to expect and work with.

**Comment by dsoprea** [comment]
Yes.. You were right, above. Not _Close()_... I meant _Release()_. Mind glitch.

**Comment by hanwen** [comment]
I'd love to take in your feedback as comments on the above code review. thanks.

**Comment by dsoprea** [comment]
That's the first thing I tried, but I don't have an account on gerrithub.

**Comment by hanwen** [comment]
there is a "sign in button top right"; it uses your Github identity. You can pick "Reviewer" as the level of access.

**Comment by hanwen** [comment]
I submitted this after extensive rework. See  https://godoc.org/github.com/hanwen/go-fuse/fs

I have some ideas on how to do an example using loopback, but I'll have to wait until later.

**Comment by dsoprea** [comment]
It's quite a bit better than it was. Thanks for looking at it. I posted just one comment about the naming of one of your examples.

When you have time, you might consider posting a reader example, and/or a reader-and-writer example. That'll be the minimum goal for the PoC of any developer.

**Comment by hanwen** [comment]
read: https://review.gerrithub.io/c/hanwen/go-fuse/+/462665/1

read/write: https://review.gerrithub.io/c/hanwen/go-fuse/+/462665/1

naming: https://review.gerrithub.io/c/hanwen/go-fuse/+/462661/1

**Comment by hanwen** [comment]
zipfs: https://review.gerrithub.io/c/hanwen/go-fuse/+/462667

**Comment by hanwen** [comment]
I've submitted these. I hope this covers your questions. Let me know if you find further unclear parts of the library.

**Comment by marco-m** [comment]
Thanks @hanwen !

**Comment by stevegt** [comment]
For anyone else who finds this more recently -- gocryptfs has been ported to the go-fuse v2 fs API, and should also be able to provide plenty of real-world example code -- ref #399.
